keps/sig-cli/kustomize-ordering.md

```diff
@@ -0,0 +1,179 @@
+- matchAnnotations:
+    some-annotation-name-2: some-annotation-value-2
+```
+
+The explicit user defined ordering using annotations will take precedence over the type based
+orderings.  Types would be used as a fallback sorting function amongst Resource with equal
+precedence in the explicit ordering.
+
+- Resources annotated with `some-annotation-name-1: some-annotation-value-1` *and* annotated
+  with `some-annotation-name-a: some-annotation-value-a` are first
+  - These Resources are sorted by type
+- Resources annotated with `some-annotation-name-1=some-annotation-value-1`
+  (without `some-annotation-name-a: some-annotation-value-a`) appear second
+  - These Resources are sorted by type
+- Resources annotated with `some-annotation-name-2=some-annotation-value-2` appear third
+  - These Resources are sorted by type
+- Resources not matching any annotation are last
+  - These Resources are sorted by type
+
+Resources matching multiple orderings (e.g. have multiple matching annotations) appear
+in the position matching the earliest label / annotation.
+
+**Note:** Throw an error if there is a selector that selects a superset of another selector 
+and it appears first.  e.g. this should throw an error because the first selector will match
+everything that the second selector does, and it will have no effect.
+
+```
+sortOrder:
+- matchAnnotations:
+    some-annotation-name-1: some-annotation-value-1
+- matchAnnotations:
+    some-annotation-name-1: some-annotation-value-1
+    some-annotation-name-a: some-label-value-a
+```
+
+### Risks and Mitigations
+
+Risk: Users build complex orderings that are hard to reason about.
+Mitigation: Good documentation and recommendations about how to keep things simple.
+
+## Graduation Criteria
+
+Use customer feedback to determine if we should support:
+
+- `LabelSelectors`
+- `AnnotationSelectorRequirement` (like `LabelSelectorRequirement` but for annotations)
+- Explicitly order "default" (e.g. Resources that don't match any of the selectors) instead of
+  them being last.
+
+
+## Docs
+
+Update Kubectl Book and Kustomize Documentation
+
+## Test plan
+
+Unit Tests for:
+
+- [ ] Resources with annotations follow the ordering override semantics
+- [ ] Resources matching multiple orderings land in the right spot
+- [ ] Throw an error if a superset selector appears after its subset (e.g. a more restrict selector appears later)
+  - This will have no effect, as everything will have already been matched
+- [ ] Resources are sorted by type as a secondary factor
+- [ ] Having multiple annotations with the same key and different values works correctly
+- [ ] Having multiple annotations with different keys and the same values works correctly
+- [ ] Empty and Null `MatchAnnotations` throw an error
+- [ ] Resources that don't appear in the ordering overrides appear last and are sorted by type
+
+### Version Skew Tests
+
+## Implementation History
+
+## Alternatives
```

Comment:
Please include a discussion of at least some alternatives, otherwise what's the point of the KEP?
(If there is only one possible option, then there is no need for a KEP doc.  If there are alternatives, then better to have the rationale for selecting this particular approach expounded by the original authors)

In particular for this doc: The obvious alternative that I think needs to be considered is just process the resources in original document order.  This is presumably the order intended by the upstream manifest author (since this is kubectl's behaviour).  We would need a bit of explanation for the kustomize corner cases like newly-inserted resources, and when multiple upstreams are merged (conceivably any order that preserves the original separtae document order(s) is fine here).

Reply:
> Please include a discussion of at least some alternatives, otherwise what's the point of the KEP?

To get other community members to come up with alternatives :)

Reply:
> The obvious alternative that I think needs to be considered is just process the resources in original document order

Would this be mutually exclusive with the sorted order, so users would need to manually put CRDs first, etc?

Reply:
> Would this be mutually exclusive with the sorted order

I don't understand...  Obviously you can't primary-sort by both document order _and_ some other (eg: type-based) order.

> users would need to manually put CRDs first, etc?

Yes, "document order" assumes that the user (or some other tool) has appropriately sorted the resources in the input document.  Again: `kubectl` works in document order, so this is already quite a common workaround/technique.